src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs

```diff
@@ -52,18 +54,49 @@ public override void ReadEnumerable(JsonTokenType tokenType, ref ReadStack state
         {
             Debug.Assert(ShouldDeserialize);
 
-            if (state.Current.KeyName == null && (state.Current.IsProcessingDictionary || state.Current.IsProcessingImmutableDictionary))
+            if (state.Current.KeyName == null && (state.Current.IsProcessingDictionary || state.Current.IsProcessingIDictionaryConstructibleOrKeyValuePair))
```

Comment:
I hope we can remove these types of checks by just letting custom converters do their own work. This is going to be buggy until then IMO - how would someone know whether to check for IsProcessingIDictionaryConstructibleOrKeyValuePair or not?